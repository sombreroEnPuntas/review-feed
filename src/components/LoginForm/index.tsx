import React, { useState, ChangeEvent, FormEvent } from 'react'
import { Box, Button, Input, Modal } from 'retro-ui'
import styled from 'styled-components'

import Error from '../Error'
import Spacer from '../Spacer'

import useLogin from '../../utils/useLogin'

const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: stretch;
`

const LoginForm = () => {
  const [username, setUsername] = useState('')
  const [password, setPassword] = useState('')

  const { error, login, logout, token } = useLogin()

  const handleLogin = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault()
    await login(username, password)
  }

  const handleLogout = () => {
    logout()
  }

  return token ? (
    <Button onClick={handleLogout}>{'Logout'}</Button>
  ) : (
    <Modal>
      <Box error={!!error}>
        <Form onSubmit={handleLogin}>
          <Input
            aria-label="username"
            autoComplete="username"
            name="username"
            onChange={(event: ChangeEvent<HTMLInputElement>) =>
              setUsername(event?.target?.value)
            }
            placeholder="username"
            required
            type="text"
            value={username}
          />
          <Spacer />
          <Input
            aria-label="password"
            autoComplete="password"
            name="password"
            onChange={(event: ChangeEvent<HTMLInputElement>) =>
              setPassword(event?.target?.value)
            }
            placeholder="password"
            required
            type="password"
            value={password}
          />
          <Error>{error ? error : '\n\n'}</Error>
          <Button>{`Login`}</Button>
        </Form>
      </Box>
    </Modal>
  )
}

export default LoginForm
