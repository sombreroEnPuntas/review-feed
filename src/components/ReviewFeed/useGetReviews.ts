import { useRef, useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { DefaultApi, ModelReview, ModelTheme } from '../../client/api'
import { Configuration } from '../../client/configuration'
import { getToken } from '../../data/login'
import { getThemes, setThemes } from '../../data/themes'
import parseAPIError from '../../utils/parseAPIError'

interface ModelThemeResponse {
  data: ModelTheme[]
}

interface ModelReviewResponse {
  data: ModelReview[]
}

const useGetReviews = () => {
  const service = useRef<DefaultApi>()

  const dispatch = useDispatch()
  const token: string = useSelector(getToken)
  const themes: ModelTheme[] = useSelector(getThemes)

  const [error, setError] = useState<string>()
  const [loading, setLoading] = useState(false)
  const [reviews, setReviews] = useState<ModelReview[]>()

  const fetchThemes = async ({
    limit,
    offset,
  }: {
    limit?: number
    offset?: number
  }) => {
    setLoading(true)
    let message = null

    try {
      // wrong types on the model :(
      const response: unknown = await service.current.apiThemesGet(
        limit,
        offset
      )
      dispatch(setThemes((response as ModelThemeResponse).data))
    } catch (e) {
      // indirect tests assert this behavior 🔍
      /* istanbul ignore next */
      message = await parseAPIError(e, dispatch)
    }

    setError(message)
    setLoading(false)
  }

  const getReviews = async ({
    limit = 10,
    offset,
    themeId,
  }: {
    limit?: number
    offset?: number
    themeId?: number
  } = {}) => {
    setLoading(true)
    let message = null

    try {
      // wrong types on the model :(
      const response: unknown = await service.current.apiReviewsGet(
        limit,
        offset,
        themeId
      )
      setError(null)
      setReviews((prevData) => [
        ...(prevData || []),
        ...(response as ModelReviewResponse).data,
      ])
    } catch (e) {
      message = await parseAPIError(e, dispatch)
    }

    setError(message)
    setLoading(false)
  }

  // instantiate the service when new tokens come only
  useEffect(() => {
    ;(async () => {
      if (!token) return

      const config: Configuration = { apiKey: `Bearer ${token}` }
      service.current = new DefaultApi(config)

      if (!themes) {
        // initializes all, since it's just text
        await fetchThemes({ limit: 100 })
        await getReviews()
      }
    })()

    return function cleanUp() {
      service.current = undefined
    }
  }, [token])

  return { error, loading, reviews, themeList: themes, getReviews }
}

export default useGetReviews
