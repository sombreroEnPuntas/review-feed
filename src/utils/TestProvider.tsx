import React from 'react'
import { Provider } from 'react-redux'
import { ThemeWrapper } from 'retro-ui'

import store from '../data'

const TestProvider = ({ children }) => (
  <ThemeWrapper>
    <Provider store={store}>{children}</Provider>
  </ThemeWrapper>
)

export default TestProvider
