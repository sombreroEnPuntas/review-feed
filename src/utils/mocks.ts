import {
  ModelJWTResponse,
  ModelAPIError,
  ModelTheme,
  ModelReview,
} from '../client/api'

export const internalError = new Error(`Wut?. SyntaxError or NetworkError`)

export const errorAPIResponse: ModelAPIError = {
  code: 418,
  message: `The resulting entity body MAY be short and stout.`,
}

export const unauthorizedAPIResponse: ModelAPIError = {
  code: 401,
  message: `You shall not pass!`,
}

export interface MockData {
  login?: ModelJWTResponse
  themes?: ModelTheme[]
  reviews?: ModelReview[]
}
export class MockService {
  data
  error

  constructor(data: MockData, error: 'internal' | 'API' | 'EXPIRED_TOKEN') {
    this.data = data
    this.error = error
  }

  private genericErrorResponse = () => {
    if (this.error === 'API') {
      throw { json: () => Promise.resolve(errorAPIResponse) }
    } else if (this.error === 'EXPIRED_TOKEN') {
      throw { json: () => Promise.resolve(unauthorizedAPIResponse) }
    } else if (this.error === 'internal') {
      throw internalError
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  loginPost = async (_username: string, _password: string) => {
    this.genericErrorResponse()

    return this.data?.login
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  apiThemesGet = async (_limit: number, _offset: number) => {
    this.genericErrorResponse()

    return { data: this.data?.themes }
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  apiReviewsGet = async (_limit: number, _offset: number, _themeId: number) => {
    this.genericErrorResponse()

    return { data: this.data?.reviews }
  }
}

export const reviewMock = {
  id: 59435909,
  created_at: '2020-08-20T11:54:52Z',
  comment: 'Good, Bad, and Ugly.',
  themes: [
    { theme_id: 1000, sentiment: 1 },
    { theme_id: 1001, sentiment: 0 },
    { theme_id: 1002, sentiment: -1 },
  ],
}

export const themeListMock = [
  { id: 1000, name: 'General' },
  { id: 1001, name: 'Other' },
  { id: 1002, name: 'Unspecific' },
]
