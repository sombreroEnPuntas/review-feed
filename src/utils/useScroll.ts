import { useState, useEffect } from 'react'

const useScroll = (callback) => {
  const [loading, setLoading] = useState(false)
  const [scroll, setScroll] = useState(0)

  function handleScroll() {
    if (
      window.innerHeight + document.documentElement.scrollTop !==
        document.documentElement.offsetHeight ||
      loading
    )
      return

    setScroll(document.documentElement.scrollTop)
    setLoading(true)
  }

  useEffect(() => {
    window.addEventListener('scroll', handleScroll)

    return function cleanUp() {
      window.removeEventListener('scroll', handleScroll)
    }
  }, [])

  useEffect(() => {
    if (!loading) return
    ;(async () => {
      await callback()
      setLoading(false)
      document.documentElement.scrollTop = scroll
    })()
  }, [loading])
}

export default useScroll
