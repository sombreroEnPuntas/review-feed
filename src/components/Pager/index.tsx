import { Button } from 'retro-ui'
import styled from 'styled-components'

import LoginForm from '../LoginForm'

const pagerColors = {
  accent: '#E4303E',
  body: '#3C3B41',
  main: '#ADACB1',
  screenBorder: '#D8D7D9',
}

const Body = styled.div`
  background-color: ${pagerColors.body};
  padding: 1em;
`

const BarContainer = styled.div`
  align-items: flex-end;
  display: flex;
  justify-content: space-between;
`
const Title = styled.h3`
  color: ${pagerColors.main};
  margin: 0 0 0.5em 0;
  max-width: max-content;
  padding: 0;
`
const StatusBar = styled.div<{ error?: boolean }>`
  ${({ theme, error }) => `
    align-items: center;
    background-color: ${pagerColors.screenBorder};
    display: flex;
    height: 20px;
    justify-content: center;
    width: 50px;

    &:before {
      background-color: ${theme.colors[error ? 'error' : 'success'].shadow};
      content: '';
      display: inline-block;
      height: 5px;
      width: 20px;
    }
  `}
`

const ScreenContainer = styled.div`
  background-color: ${pagerColors.screenBorder};
  display: flex;
  flex-direction: column;
`
const Screen = styled.div<{ error?: boolean }>`
  ${({ theme, error }) => `
    background-color: ${theme.colors[error ? 'error' : 'success'].background};
    box-shadow: \
      4px 0 ${pagerColors.body}, \
      0 -4px ${pagerColors.body}, \
      -4px 0 ${pagerColors.body}, \
      0 4px ${pagerColors.body};
    color: ${theme.colors[error ? 'error' : 'success'].outline};
    height: 4em;
    margin: 1em;
    overflow: auto;
    padding: 0.25em;
  `}
`

const Actions = styled.div`
  display: flex;
  justify-content: space-between;
  margin-top: 0.5em;
`
const ActionsGroup = styled.div`
  display: flex;
`
const MainCTA = styled(Button)`
  background-color: ${pagerColors.main};
  box-shadow: none;
`
export const AccentCTA = styled(Button)`
  background-color: ${pagerColors.accent};
  box-shadow: none;
`

export interface Props {
  error?: boolean
  message: string
}
const Pager = ({ error, message }: Props) => (
  <Body>
    <BarContainer>
      <Title>{'review-feed'}</Title>
      <StatusBar error={!!error} />
    </BarContainer>
    <ScreenContainer>
      <Screen error={!!error}>{message}</Screen>
    </ScreenContainer>
    <Actions>
      <ActionsGroup>
        <MainCTA>{'Filter'}</MainCTA>
        <MainCTA>{'Reset'}</MainCTA>
      </ActionsGroup>
      <LoginForm />
    </Actions>
  </Body>
)

export default Pager
