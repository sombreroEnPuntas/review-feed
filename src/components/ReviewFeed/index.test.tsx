import { render } from '@testing-library/react'
import React from 'react'

import TestedComponent from '.'

// Deps
import useGetReviews from '../../utils/useGetReviews'
import useScroll from '../../utils/useScroll'
import LoginForm from '../LoginForm'
import { AccentCTA } from '../Pager'

// Utils
import { reviewMock, themeListMock } from '../../utils/mocks'
import TestProvider from '../../utils/TestProvider'

// Mocks
jest.mock('../../utils/useGetReviews')
jest.mock('../../utils/useScroll')
jest.mock('../LoginForm')

// Mock data
const setMock = ({ error, loading, reviews, themeList, getReviews }) => {
  ;(useGetReviews as jest.Mock).mockImplementation(() => ({
    error,
    loading,
    reviews,
    themeList,
    getReviews,
  }))
  ;(LoginForm as jest.Mock).mockImplementation(() => (
    <AccentCTA>{'Logout'}</AccentCTA>
  ))
}

const stageTest = ({
  error,
  loading,
  reviews = null,
  themeList = null,
  getReviews = jest.fn(),
}) => {
  setMock({ error, loading, reviews, themeList, getReviews })

  return render(
    <TestProvider>
      <TestedComponent />
    </TestProvider>
  )
}

describe('ReviewFeed', () => {
  afterEach(() => {
    // much force-casting, such type-safe, so dark side
    ;(useGetReviews as jest.Mock).mockClear()
    ;(useScroll as jest.Mock).mockClear()
    ;(LoginForm as jest.Mock).mockClear()
  })
  afterAll(() => {
    jest.resetAllMocks() // clean .mock
  })

  it(`renders a successful feed`, async () => {
    const testData = {
      error: null,
      loading: false,
      reviews: [reviewMock],
      themeList: [...themeListMock],
    }
    const { container, getByText } = stageTest(testData)

    const appState = getByText(/Listing all themes./)
    const review = getByText(/Good, Bad, and Ugly./)

    expect(appState).toBeInTheDocument()
    expect(review).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(`renders a loading feed`, async () => {
    const testData = {
      error: null,
      loading: true,
    }
    const { container, getByText } = stageTest(testData)

    const appState = getByText(/Loading.../)

    expect(appState).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(`renders a failed feed`, async () => {
    const testData = {
      error: `418: The resulting entity body MAY be short and stout.`,
      loading: false,
    }
    const { container, getByText } = stageTest(testData)

    const appState = getByText(
      /418: The resulting entity body MAY be short and stout./
    )

    expect(appState).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(`renders a failed loading feed`, async () => {
    const testData = {
      error: `418: The resulting entity body MAY be short and stout.`,
      loading: true,
    }
    const { container, getByText } = stageTest(testData)

    const appState = getByText(/Loading.../)

    expect(appState).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(`initializes scroll listener`, async () => {
    const testData = {
      error: null,
      loading: false,
    }
    stageTest(testData)

    expect(useScroll).toHaveBeenCalled()
  })
})
