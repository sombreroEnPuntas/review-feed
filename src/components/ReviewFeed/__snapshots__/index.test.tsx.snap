// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewFeed renders a failed feed 1`] = `
.c1 {
  background-color: #3C3B41;
  padding: 1em;
}

.c2 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  color: #ADACB1;
  margin: 0 0 0.5em 0;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  padding: 0;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #D8D7D9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50px;
}

.c4:before {
  background-color: #e76e55;
  content: '';
  display: inline-block;
  height: 5px;
  width: 20px;
}

.c5 {
  background-color: #D8D7D9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  background-color: #ffd7cf;
  box-shadow: 4px 0 #3C3B41,0 -4px #3C3B41,-4px 0 #3C3B41,0 4px #3C3B41;
  color: #ce372b;
  height: 4em;
  margin: 1em;
  overflow: auto;
  padding: 0.25em;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: 0.5em;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #ADACB1;
  box-shadow: none;
}

.c10 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #E4303E;
  box-shadow: none;
}

.c0 {
  margin: auto;
  max-width: 450px;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1;
}

.c11 {
  margin: auto;
  max-width: 450px;
  padding: 0 0.5em;
  position: relative;
  z-index: 0;
}

<div>
  <header
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h3
          class="c3"
        >
          review-feed
        </h3>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          418: The resulting entity body MAY be short and stout.
        </div>
      </div>
      <div
        class="c7"
      >
        <div
          class="c8"
        >
          <button
            class="sc-AxiKw c9"
          >
            Filter
          </button>
          <button
            class="sc-AxiKw c9"
          >
            Reset
          </button>
        </div>
        <button
          class="sc-AxiKw c10"
        >
          Logout
        </button>
      </div>
    </div>
  </header>
  <section
    class="c11"
  />
</div>
`;

exports[`ReviewFeed renders a failed loading feed 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #4d525666;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 0;
  overflow: auto;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c2 {
  background-color: #3C3B41;
  padding: 1em;
}

.c3 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  color: #ADACB1;
  margin: 0 0 0.5em 0;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  padding: 0;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #D8D7D9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50px;
}

.c5:before {
  background-color: #e76e55;
  content: '';
  display: inline-block;
  height: 5px;
  width: 20px;
}

.c6 {
  background-color: #D8D7D9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c7 {
  background-color: #ffd7cf;
  box-shadow: 4px 0 #3C3B41,0 -4px #3C3B41,-4px 0 #3C3B41,0 4px #3C3B41;
  color: #ce372b;
  height: 4em;
  margin: 1em;
  overflow: auto;
  padding: 0.25em;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: 0.5em;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c10 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #ADACB1;
  box-shadow: none;
}

.c11 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #E4303E;
  box-shadow: none;
}

.c1 {
  margin: auto;
  max-width: 450px;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1;
}

.c12 {
  margin: auto;
  max-width: 450px;
  padding: 0 0.5em;
  position: relative;
  z-index: 0;
}

<div>
  <div
    class="c0"
  />
  <header
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <h3
          class="c4"
        >
          review-feed
        </h3>
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          Loading...
        </div>
      </div>
      <div
        class="c8"
      >
        <div
          class="c9"
        >
          <button
            class="sc-AxiKw c10"
          >
            Filter
          </button>
          <button
            class="sc-AxiKw c10"
          >
            Reset
          </button>
        </div>
        <button
          class="sc-AxiKw c11"
        >
          Logout
        </button>
      </div>
    </div>
  </header>
  <section
    class="c12"
  />
</div>
`;

exports[`ReviewFeed renders a loading feed 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #4d525666;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 0;
  overflow: auto;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c2 {
  background-color: #3C3B41;
  padding: 1em;
}

.c3 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  color: #ADACB1;
  margin: 0 0 0.5em 0;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  padding: 0;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #D8D7D9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50px;
}

.c5:before {
  background-color: #92cc41;
  content: '';
  display: inline-block;
  height: 5px;
  width: 20px;
}

.c6 {
  background-color: #D8D7D9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c7 {
  background-color: #c9d68f;
  box-shadow: 4px 0 #3C3B41,0 -4px #3C3B41,-4px 0 #3C3B41,0 4px #3C3B41;
  color: #76c442;
  height: 4em;
  margin: 1em;
  overflow: auto;
  padding: 0.25em;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: 0.5em;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c10 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #ADACB1;
  box-shadow: none;
}

.c11 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #E4303E;
  box-shadow: none;
}

.c1 {
  margin: auto;
  max-width: 450px;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1;
}

.c12 {
  margin: auto;
  max-width: 450px;
  padding: 0 0.5em;
  position: relative;
  z-index: 0;
}

<div>
  <div
    class="c0"
  />
  <header
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <h3
          class="c4"
        >
          review-feed
        </h3>
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
      >
        <div
          class="c7"
        >
          Loading...
        </div>
      </div>
      <div
        class="c8"
      >
        <div
          class="c9"
        >
          <button
            class="sc-AxiKw c10"
          >
            Filter
          </button>
          <button
            class="sc-AxiKw c10"
          >
            Reset
          </button>
        </div>
        <button
          class="sc-AxiKw c11"
        >
          Logout
        </button>
      </div>
    </div>
  </header>
  <section
    class="c12"
  />
</div>
`;

exports[`ReviewFeed renders a successful feed 1`] = `
.c13 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  display: block;
  margin: 0.5em;
  max-width: 450px;
  overflow-x: auto;
  padding: 0.5em;
  white-space: pre-wrap;
}

.c14 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #212529;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  line-height: calc(2 * 1.5em);
}

.c1 {
  background-color: #3C3B41;
  padding: 1em;
}

.c2 {
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  color: #ADACB1;
  margin: 0 0 0.5em 0;
  max-width: -webkit-max-content;
  max-width: -moz-max-content;
  max-width: max-content;
  padding: 0;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #D8D7D9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50px;
}

.c4:before {
  background-color: #92cc41;
  content: '';
  display: inline-block;
  height: 5px;
  width: 20px;
}

.c5 {
  background-color: #D8D7D9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  background-color: #c9d68f;
  box-shadow: 4px 0 #3C3B41,0 -4px #3C3B41,-4px 0 #3C3B41,0 4px #3C3B41;
  color: #76c442;
  height: 4em;
  margin: 1em;
  overflow: auto;
  padding: 0.25em;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: 0.5em;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #ADACB1;
  box-shadow: none;
}

.c10 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #E4303E;
  box-shadow: none;
}

.c12 {
  background-color: white;
  font-size: 0.5em;
  left: 2em;
  line-height: 1.5em;
  padding: 0 0.5em;
  position: relative;
  top: 1.2em;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.c15 {
  display: block;
  width: 100%;
}

.c16 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c17 {
  background-color: #c9d68f;
  margin: 0.5em;
  padding: 0 0.5em;
  white-space: nowrap;
}

.c18 {
  background-color: #E9C46A;
  margin: 0.5em;
  padding: 0 0.5em;
  white-space: nowrap;
}

.c19 {
  background-color: #ffd7cf;
  margin: 0.5em;
  padding: 0 0.5em;
  white-space: nowrap;
}

.c20 {
  margin: 0 1em;
}

.c0 {
  margin: auto;
  max-width: 450px;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1;
}

.c11 {
  margin: auto;
  max-width: 450px;
  padding: 0 0.5em;
  position: relative;
  z-index: 0;
}

<div>
  <header
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h3
          class="c3"
        >
          review-feed
        </h3>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          Listing all themes.
        </div>
      </div>
      <div
        class="c7"
      >
        <div
          class="c8"
        >
          <button
            class="sc-AxiKw c9"
          >
            Filter
          </button>
          <button
            class="sc-AxiKw c9"
          >
            Reset
          </button>
        </div>
        <button
          class="sc-AxiKw c10"
        >
          Logout
        </button>
      </div>
    </div>
  </header>
  <section
    class="c11"
  >
    <article>
      <div
        class="c12"
      >
        8/20/2020
      </div>
      <pre
        class="c13"
        type="black"
      >
        <code
          class="c14"
          type="black"
        >
          <div
            class="c15"
          >
            <div
              class="c16"
            >
              <span
                class="c17"
              >
                General
              </span>
              <span
                class="c18"
              >
                Other
              </span>
              <span
                class="c19"
              >
                Unspecific
              </span>
            </div>
            <hr />
            <div
              class="c20"
            >
              Good, Bad, and Ugly.
            </div>
          </div>
        </code>
      </pre>
    </article>
  </section>
</div>
`;
